File: internal/seedphrase/seedphrase.go
// Package seedphrase turns the secret that groups a person's own instances
// into twelve words, and back.
//
// This is the whole of what somebody copies from one KnightLoader to
// another (jdp, 2026-08-27: "eine zeichenfolge oder eine Seed phrase ...
// die man dann in allen anderen Instanzen einfügen kann"). It carries the
// secret and nothing else - not an address, because the relay's address is
// a compiled-in constant, which is exactly what lets the phrase stay this
// short and why there is no login step anywhere in the flow. See
// docs/superpowers/specs/2026-08-27-public-relay-seed-phrase-design.md.
//
// # Why the BIP39 wordlist
//
// Not because of cryptocurrency, and the UI must never call this a wallet
// seed. The list is reused because of one property that is expensive to
// reproduce and easy to get wrong: its 2048 words were chosen so that no
// two share their first four letters, none are near-homophones, and none
// carry accents. That is what makes a phrase safe to read aloud down a
// phone line and safe to type on a mobile keyboard - the two things this
// format exists for. The word file is the official list, embedded rather
// than fetched, and its content is pinned by TestWordlistIsTheOfficialOne.
//
// # Why 128 bits
//
// An earlier draft of the relay design specified 256. That is 24 words,
// and the difference between a phrase somebody will actually type on a
// phone and one they will give up on. At 2^128 guessing is not a threat
// model any rate limit needs to help with, so the extra 12 words buy
// nothing real and cost the feature its usability.
//
// The BIP39 checksum rides along for free and earns its place: a mistyped
// word is caught here, with the offending word named, instead of surfacing
// later as a connection that silently never finds its sibling.
package seedphrase

import (
	"crypto/rand"
	"crypto/sha256"
	_ "embed"
	"errors"
	"fmt"
	"strings"
)

//go:embed english.txt
var wordlistFile string

const (
	// SecretLen is the secret's size in bytes. 128 bits - see the package
	// comment for why not more.
	SecretLen = 16
	// WordCount is what SecretLen encodes to: 128 bits of secret plus a
	// 4-bit checksum, split into 11-bit groups.
	WordCount = 12

	bitsPerWord  = 11
	checksumBits = SecretLen * 8 / 32 // BIP39's own rule: one bit per 32 bits of entropy
)

var (
	words     []string
	wordIndex map[string]int
)

func init() {
	words = strings.Fields(wordlistFile)
	if len(words) != 1<<bitsPerWord {
		// A wordlist of the wrong length cannot encode 11 bits per word, so
		// every phrase this package produced would be undecodable. Failing
		// at init is the only honest response: there is no degraded mode.
		panic(fmt.Sprintf("seedphrase: wordlist has %d words, need %d", len(words), 1<<bitsPerWord))
	}
	wordIndex = make(map[string]int, len(words))
	for i, w := range words {
		wordIndex[w] = i
	}
}

// New mints a fresh secret and returns it with its phrase. The secret is
// what the relay sees; the phrase is what a person copies.
func New() (secret []byte, phrase string, err error) {
	secret = make([]byte, SecretLen)
	if _, err := rand.Read(secret); err != nil {
		// crypto/rand failing is a broken machine, not a path to recover
		// from - and silently returning a weak secret here would hand out a
		// phrase that looks exactly as trustworthy as a real one.
		return nil, "", fmt.Errorf("seedphrase: reading randomness: %w", err)
	}
	phrase, err = Encode(secret)
	if err != nil {
		return nil, "", err
	}
	return secret, phrase, nil
}

// Encode renders a secret as its phrase. Used both when a group is first
// created and every time the phrase is shown again later (which is gated
// behind re-entering the instance password - see the spec).
func Encode(secret []byte) (string, error) {
	if len(secret) != SecretLen {
		return "", fmt.Errorf("seedphrase: secret is %d bytes, need %d", len(secret), SecretLen)
	}
	sum := sha256.Sum256(secret)
	// secret || first checksumBits of sha256(secret), read as one bit
	// string. One spare byte is enough for any checksum this size.
	full := make([]byte, 0, SecretLen+1)
	full = append(full, secret...)
	full = append(full, sum[0])

	out := make([]string, WordCount)
	for i := range out {
		out[i] = words[bitsAt(full, i*bitsPerWord, bitsPerWord)]
	}
	return strings.Join(out, " "), nil
}

// ErrChecksum is returned when every word is real but the phrase as a whole
// does not check out - the signature of a typo that happened to land on
// another valid word, or of two words swapped.
var ErrChecksum = errors.New("seedphrase: the phrase is not valid - check for a mistyped or swapped word")

// Decode parses a phrase back into its secret.
//
// Input is normalised first, because this arrives from a paste or from
// somebody typing what they were read over the phone: case is ignored, and
// any run of whitespace (including the newlines a copied block brings with
// it) counts as one separator.
func Decode(phrase string) ([]byte, error) {
	got := strings.Fields(strings.ToLower(strings.TrimSpace(phrase)))
	if len(got) != WordCount {
		return nil, fmt.Errorf("seedphrase: the phrase has %d words, it needs %d", len(got), WordCount)
	}

	full := make([]byte, SecretLen+1)
	for i, w := range got {
		idx, ok := wordIndex[w]
		if !ok {
			// Naming the word and its position is the whole point: "word 7
			// (\"recieve\") is not in the list" is a fixable message,
			// "invalid phrase" is not.
			return nil, fmt.Errorf("seedphrase: word %d (%q) is not one of the accepted words", i+1, w)
		}
		setBits(full, i*bitsPerWord, bitsPerWord, idx)
	}

	secret := full[:SecretLen]
	sum := sha256.Sum256(secret)
	// Compare only the checksum bits, which sit immediately after the
	// secret; the rest of the trailing byte is padding this format never
	// wrote to.
	want := bitsAt(sum[:1], 0, checksumBits)
	have := bitsAt(full, SecretLen*8, checksumBits)
	if want != have {
		return nil, ErrChecksum
	}
	return secret, nil
}

// bitsAt reads count bits from b starting at bit offset, most significant
// bit first, and returns them as an integer. Written plainly rather than
// cleverly: it runs twelve times per phrase, so there is nothing to gain
// from packing this into shifts nobody can check by eye.
func bitsAt(b []byte, offset, count int) int {
	v := 0
	for i := 0; i < count; i++ {
		bit := offset + i
		v <<= 1
		if b[bit/8]&(1<<(7-bit%8)) != 0 {
			v |= 1
		}
	}
	return v
}

// setBits is bitsAt's inverse: writes the low count bits of v into b at bit
// offset, most significant bit first. Assumes the target bits start clear,
// which they do - callers hand it a freshly made slice.
func setBits(b []byte, offset, count, v int) {
	for i := 0; i < count; i++ {
		if v&(1<<(count-1-i)) != 0 {
			bit := offset + i
			b[bit/8] |= 1 << (7 - bit%8)
		}
	}
}

File: mobile/src/api/seedphrase.ts
// The phone's half of the connection phrase.
//
// It decodes locally rather than asking a server, because the whole point of
// the phrase is the case where this phone cannot reach any instance yet -
// there is nobody to ask. A port of internal/seedphrase, deliberately line
// for line: the two must agree on every one of the 2048 words and on the bit
// packing, or the key derived here quietly differs from the server's and the
// symptom is "the relay never connects" with nothing in any log.
//
// The UI must never call this a wallet seed. It is a Verbindungsphrase.

import { sha256, toHex, utf8 } from './sha256';
import { WORDS } from './wordlist';

export const SECRET_LEN = 16;
export const WORD_COUNT = 12;
const BITS_PER_WORD = 11;
const CHECKSUM_BITS = (SECRET_LEN * 8) / 32; // BIP39's own rule

// relay.DefaultRelayURL, compiled in on the server side for the same reason
// it is a constant here: it is what keeps a phrase twelve words instead of a
// URL plus a key.
export const DEFAULT_RELAY_URL = 'wss://relay.knightloader.app/relay/connect';

// relay.keyDomain. Changing this string orphans every phrase in existence.
const KEY_DOMAIN = 'knightloader/relay/group-key/v1';

const INDEX = new Map<string, number>(WORDS.map((w, i) => [w, i]));

/** Why a phrase was refused, so the caller can say so in the reader's language. */
export type PhraseProblem =
  | { reason: 'word_count'; count: number }
  | { reason: 'unknown_word'; word: string; position: number }
  | { reason: 'checksum' };

export class PhraseError extends Error {
  problem: PhraseProblem;
  constructor(problem: PhraseProblem) {
    super(problem.reason);
    this.name = 'PhraseError';
    this.problem = problem;
  }
}

/** bitsAt reads count bits from b starting at bit offset, most significant first. */
function bitsAt(b: Uint8Array, offset: number, count: number): number {
  let v = 0;
  for (let i = 0; i < count; i++) {
    const bit = offset + i;
    v <<= 1;
    if (b[bit >> 3] & (1 << (7 - (bit & 7)))) v |= 1;
  }
  return v;
}

/** setBits is bitsAt's inverse; the target bits start clear. */
function setBits(b: Uint8Array, offset: number, count: number, v: number): void {
  for (let i = 0; i < count; i++) {
    if (v & (1 << (count - 1 - i))) {
      const bit = offset + i;
      b[bit >> 3] |= 1 << (7 - (bit & 7));
    }
  }
}

/**
 * decodePhrase parses twelve words back into the secret they carry.
 *
 * Input is normalised first, because it arrives from a paste, a QR scan, or
 * somebody typing what was read to them: case is ignored and any run of
 * whitespace counts as one separator.
 *
 * Throws PhraseError, never a bare string, so the screen can pick a
 * translated sentence instead of showing one this file chose.
 */
export function decodePhrase(phrase: string): Uint8Array {
  const got = phrase.trim().toLowerCase().split(/\s+/).filter(Boolean);
  if (got.length !== WORD_COUNT) {
    throw new PhraseError({ reason: 'word_count', count: got.length });
  }

  const full = new Uint8Array(SECRET_LEN + 1);
  for (let i = 0; i < got.length; i++) {
    const idx = INDEX.get(got[i]);
    if (idx === undefined) {
      // Naming the word and its position is the point: bisecting a
      // twelve-word phrase by hand is not a thing to ask of anybody.
      throw new PhraseError({ reason: 'unknown_word', word: got[i], position: i + 1 });
    }
    setBits(full, i * BITS_PER_WORD, BITS_PER_WORD, idx);
  }

  const secret = full.slice(0, SECRET_LEN);
  const sum = sha256(secret);
  const want = bitsAt(sum.slice(0, 1), 0, CHECKSUM_BITS);
  const have = bitsAt(full, SECRET_LEN * 8, CHECKSUM_BITS);
  if (want !== have) throw new PhraseError({ reason: 'checksum' });
  return secret;
}

/**
 * deriveKey turns the secret into what the relay is actually told.
 *
 * The secret never travels. The relay matches connections presenting the same
 * derived key and cannot work backwards to the words, which is what lets
 * somebody else run the relay without being able to reconstruct anybody's
 * phrase - including us.
 */
export function deriveKey(secret: Uint8Array): string {
  const domain = utf8(KEY_DOMAIN);
  const buf = new Uint8Array(domain.length + secret.length);
  buf.set(domain);
  buf.set(secret, domain.length);
  return toHex(sha256(buf));
}

/** keyFromPhrase is the whole journey, for the one caller that wants it. */
export function keyFromPhrase(phrase: string): string {
  return deriveKey(decodePhrase(phrase));
}

/**
 * FRAME_KEY_DOMAIN is the second domain over the same secret, mirroring
 * internal/relay/key.go's frameDomain. Both strings must match byte for byte
 * or this phone joins a group it cannot talk to.
 */
const FRAME_KEY_DOMAIN = 'knightloader/relay/frame-key/v1';

/**
 * deriveFrameKey returns the 32-byte key that seals proxy frames, mirroring
 * relay.DeriveFrameKey.
 *
 * The separate domain is the entire point and is worth restating here, where
 * somebody reading only the phone's half will see it: the relay is HANDED
 * deriveKey's output in every hello frame. A frame key derived from that
 * value, or under the same domain, would be a key the relay already holds,
 * and the encryption would protect nothing from the one party it is aimed at.
 */
export function deriveFrameKey(secret: Uint8Array): Uint8Array {
  const domain = utf8(FRAME_KEY_DOMAIN);
  const buf = new Uint8Array(domain.length + secret.length);
  buf.set(domain);
  buf.set(secret, domain.length);
  return sha256(buf);
}

/** frameKeyFromPhrase is deriveFrameKey's counterpart to keyFromPhrase. */
export function frameKeyFromPhrase(phrase: string): Uint8Array {
  return deriveFrameKey(decodePhrase(phrase));
}
